### PR TITLE
Python Support

### PR DESCRIPTION
**Pull request by mr2rm**
Hi,
Recently after a long time away from competitive programming, I'v decided solving some programming challenges everyday. So I searched for the available tools which I can use to do it better and have more fun. Finally I've found the `competitive-programming-helper` useful to test my solutions.

Yesterday, I've tried `competitive-programming-helper` in CodeForces contest and I had to solve problems only with C and C++. Since I'm a Python lover (of course beside C++ and JavaScript) and I use it a lot to solve my daily challenges, I decided to equip it to Python.

So because it's the starting point of support multi languages, I made some fundamental changes to take effect. Here are my changelogs: 

- [x] Add a `QuickPick` to get the user-selected language (C, C++ and Python)
- [x] Add language configuration to handle multi languages
- [x] Some changes to introduce `.py` as a valid extension (e.g. keybindings, source creator, test runners and etc.)
- [x] Implement Python compilation process
- [x] Update test case runners to support Python
- [x] Some other changes and refactors to handle multilanguage support

Any feedback and response would be appreciated. Also I hope these changes will be useful and could help everybody who wants to improve his problem solving paradigm.

### CODE REVIEW

**Comment by agrawal-d** [comment]
@mr2rm Thanks a lot for the PR! I'm very happy to see contributions from others :smile: 
Before I review this, let me try to fix these rebase issues, it appears some older commits are becoming a part of this PR.
Once again, thanks for working on this!

**Comment by mr2rm** [comment]
@agrawal-d Don't mention it :))
I had some other ideas to improve it. I'll work on them in near future.
Oops! What's wrong with it?

**Comment by agrawal-d** [comment]
Okay, it appears you did not work on the latest master. Can you please rebase onto the latest master and fix any merge conflicts?
You would do something like:
```bash
$ git fetch origin

$ git rebase origin/master
```
Now you will get come merge conflicts. Fix each of the conflicted file, after fixing the conflict, use 

```
$ git add filename

```

After doing this for all the files, do
```
$ git rebase --continue
```

Then, to push them to the PR, do

```
git push mr2rm python-support
```

This will remove the unnecessary extra commits. If you are uncomfortable with git, I'll be happy to do it for you.

**Comment by mr2rm** [comment]
> Okay, it appears you did not work on the latest master. Can you please rebase onto the latest master and fix any merge conflicts?
> You would do something like:
> 
> ```shell
> $ git fetch origin
> 
> $ git rebase origin/master
> ```
> 
> Now you will get come merge conflicts. Fix each of the conflicted file, after fixing the conflict, use
> 
> ```
> $ git add filename
> ```
> 
> After doing this for all the files, do
> 
> ```
> $ git rebase --continue
> ```
> 
> Then, to push them to the PR, do
> 
> ```
> git push mr2rm python-support
> ```
> 
> This will remove the unnecessary extra commits. If you are uncomfortable with git, I'll be happy to do it for you.

I've already merged `upstream` into own `origin/master`

**Comment by mr2rm** [comment]
> Okay, it appears you did not work on the latest master. Can you please rebase onto the latest master and fix any merge conflicts?
> You would do something like:
> 
> ```shell
> $ git fetch origin
> 
> $ git rebase origin/master
> ```
> 
> Now you will get come merge conflicts. Fix each of the conflicted file, after fixing the conflict, use
> 
> ```
> $ git add filename
> ```
> 
> After doing this for all the files, do
> 
> ```
> $ git rebase --continue
> ```
> 
> Then, to push them to the PR, do
> 
> ```
> git push mr2rm python-support --force
> ```
> 
> This will remove the unnecessary extra commits.
> If you are uncomfortable with git, I'll be happy to do it for you.

I've already merged `upstream` into `master` and `python-support` in https://github.com/agrawal-d/competitive-programming-helper/pull/18/commits/2bbdbb0efc65bde352794561d895c56ce8f3c20b and https://github.com/agrawal-d/competitive-programming-helper/pull/18/commits/b94742ed9a331c6f9ac0b62fac62ab5b92ca5ac3 commits.
I'm ok with it. I don't know how this thing happened.

**Comment by agrawal-d** [comment]
>I've already merged

That's exactly the problem - in this codebase, we use rebase, so if you merge, it will create extra commits. Don't worry, it's not a big deal. Let me know if you face any problems fixing this.

**Comment by agrawal-d** [comment]
I think you should also delete those two merge commits - b94742ed9a331c6f9ac0b62fac62ab5b92ca5ac3 and 2bbdbb0efc65bde352794561d895c56ce8f3c20b.

**Comment by mr2rm** [comment]
> I think you should also delete those two merge commits - [b94742e](https://github.com/agrawal-d/competitive-programming-helper/commit/b94742ed9a331c6f9ac0b62fac62ab5b92ca5ac3) and [2bbdbb0](https://github.com/agrawal-d/competitive-programming-helper/commit/2bbdbb0efc65bde352794561d895c56ce8f3c20b).
> 
> (using [interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)).

I've deleted those to commits both from `master`  and `python-support`.

> ```shell
> $ git fetch origin
> ```
Do you mean `upstream`?

**Comment by agrawal-d** [comment]
> Do you mean upstream?

Whatever's the name of the branch that points to this repo. Yeah, most likely `upstream`.

**Comment by mr2rm** [comment]
> > Do you mean upstream?
> 
> Whatever's the name of the branch that points to this repo. Yeah, most likely `upstream`.

I have 2 remotes, `origin` which is the forked repo and `upstream` which is original repo. I'm not sure what do you mean `origin`? Is it `origin` or `upstream`.
To update origin I need to `fetch upstream` and then `merge upstream/master` into `master`. Then I can rebase. Is it ok?

**Comment by mr2rm** [comment]
@agrawal-d Rebase issues has just resolved and it's ready to review. ‌I think everything is OK. Tell me if there was any other problem.

File: compileFile.js [hunk 1]
```diff
@@ -2,34 +2,40 @@ const { spawn } = require("child_process");
...
+        flags = ['-m', 'compileall', '-b', filePath, ...flags];
+    else {
+        // type of compiler ( g++ or gcc )
+        const outputLocation = locationHelper.getBinLocation(language, filePath);
+        flags = [filePath, "-o", outputLocation, ...flags];
+    }
+    return flags
+}
 
 // compile the file and return a promise with the result/error
```
Comment (agrawal-d) [inline compileFile.js]: Let's make this a JSDoc, ie a comment with ```/* ... */```.

File: compileFile.js [hunk 2]
```diff
@@ -2,34 +2,40 @@ const { spawn } = require("child_process");
...
+    const ext = config.extensions[language];
+    let flags = preferences().get("compilationFlags" + ext).split(" ");
+    if (flags[0] === "")
+        flags = [];
+
+    if (language === 'Python')
+        flags = ['-m', 'compileall', '-b', filePath, ...flags];
+    else {
+        // type of compiler ( g++ or gcc )
+        const outputLocation = locationHelper.getBinLocation(language, filePath);
```
Comment (agrawal-d) [inline compileFile.js]: Let's make `g++` as `gpp` instead, here, and in other places. In the past I've used both, but using only one will help in maintaining consistency.

File: compileFile.js [hunk 3]
```diff
@@ -2,34 +2,40 @@ const { spawn } = require("child_process");
...
+ * @param {string} filePath complete path to .c, .cpp or .py file
+ */
+function getFlags(language, filePath) {
+    const ext = config.extensions[language];
+    let flags = preferences().get("compilationFlags" + ext).split(" ");
+    if (flags[0] === "")
+        flags = [];
+
+    if (language === 'Python')
+        flags = ['-m', 'compileall', '-b', filePath, ...flags];
```
Comment (agrawal-d) [inline compileFile.js]: Can you make it so that compiled bytecode, ie the `.pyc` file, is at `outputLocation` instead of the working directory?

Comment (agrawal-d) [inline compileFile.js]: What is the need for the `-b` flag here? In-fact, with `python2`, it causes an error.

File: config.js [hunk 4]
```diff
@@ -1,5 +1,15 @@
...
+  port: 27121,
+  extensions: {
+    "C": "c",
+    "C++": "cpp",
+    "Python": "py"
+  },
+  compilers: {
+    "C": "gcc",
+    "C++": "g++",
+    "Python": "python"
```
Comment (agrawal-d) [inline config.js]: Would `python3` be better? In my Linux PC, `python` points to python2 and to use Python 3, I had to use `python3`. My guess is that if Python 3 is installed, then `python3` will definitely point to Python 3. I'm not sure since I have little experience with Python.

To reduce confusion, in the settings and other user-facing screens, we can say, "Compilation flags for Python 3" specifically.

File: extension.js [hunk 5]
```diff
@@ -14,9 +14,42 @@ const EventEmitter = require('events');
...
+    .toLowerCase();
+  return fileExtension;
+}
+
+/**
+ * Get language based on file extension
+ * @param {string} extension c, cpp or py
+ */
+function getLanguage(extension) {
+  for (const [lang, ext] of Object.entries(config.extensions))
```
Comment (agrawal-d) [inline extension.js]: Let's throw an error here if no matching extension is found. Although it will likely never happen, because we are aleady checking it in line 428.

File: extension.js [hunk 6]
```diff
@@ -297,9 +333,9 @@ function appendProblemURLToFile(problemURL, callback) {
 
 /**
  * show dialog box for actions such as downloading testcases and generating testcase file manually
- * @param {any} filepath path to the active source code document
+ * @param {any} filePath path to the active source code document
  */
-function testCasesHelper(filepath) {
+function testCasesHelper(filePath) {
```
Comment (agrawal-d) [inline extension.js]: This is a good change. `filePath` is better than `filepath`. :smile:

File: extension.js [hunk 7]
```diff
@@ -377,19 +413,25 @@ async function executePrimaryTask(context) {
...
+  const extListPresentation = `${
+    extList.slice(0, -1).join(', ')
+  } or ${
+    extList.slice(-1)[0]
+  }`;
+
+  if (!validExtensions.includes(fileExtension)) {
     vscode.window.showInformationMessage(
-      "Active file must be have a .c or .cpp extension"
+      `Active file must be have a ${extListPresentation} extension`
```
Comment (agrawal-d) [inline extension.js]: nit: You can add a period `.` at the end.

File: extension.js [hunk 8]
```diff
@@ -419,21 +461,26 @@ async function executePrimaryTask(context) {
...
     let stdoutlen = 0;
-    let spawned_process = spawn(locationHelper.getBinLocation(filepath), {
-      timeout: 10000
-    });
+    const binPath = locationHelper.getBinLocation(language, filePath)
+    let options = {timeout: 10000}
+    
+    let spawned_process = (language === 'Python'
+      ? spawn(config.compilers[language], [binPath], options)
+      : spawn(binPath, {timeout: 10000})
```
Comment (agrawal-d) [inline extension.js]: You can use `options` here too.

File: locationHelper.js [hunk 9]
```diff
@@ -3,24 +3,29 @@ const path = require("path");
 const preferences = require("./preferencesHelper");
 
 /**
- * @param filepath complete path to .cpp file
+ * Get compiled file
+ * @param language programming language
+ * @param filePath complete path to .c, .cpp or .py file
  */
-function getBinLocation(filepath) {
+function getBinLocation(language, filePath) {
```
Comment (agrawal-d) [inline locationHelper.js]: Maybe we can save the `.pyc` file as `.bin` file too. I don't changing the extension should confuse the Python interpreter. In this way, we can have a common location for all compiled binaries.

File: config.js (hunk 4 above)
Comment (agrawal-d) [reply config.js]: Also, can you add a small checking mechanism to see if python3 is installed, and if not, show an error to the user like "`python3` does not point to a valid Python installation" or something like that?

File: extension.js [hunk 10]
```diff
@@ -98,21 +131,24 @@ function createLayout() {
...
-      let spawned_process = spawn(locationHelper.getBinLocation(filepath), {
-        timeout: 10000
-      });
-      spawnStack.push(spawned_process);
+      const binPath = locationHelper.getBinLocation(language, filePath)
+      let options = {timeout: 10000}
 
+      let spawned_process = (language === 'Python'
+        ? spawn(config.compilers[language], [binPath], options)
+        : spawn(binPath, {timeout: 10000})
```
Comment (agrawal-d) [inline extension.js]: You can use `options` here too.

File: compileFile.js (hunk 1 above)
Comment (mr2rm) [reply compileFile.js]: Do you mean comment about what each flag does?

File: compileFile.js (hunk 2 above)
Comment (mr2rm) [reply compileFile.js]: I didn't understand what you mean. Do you want encrypt something?

Comment (mr2rm) [reply compileFile.js]: Where did you use `gpp` in the codebase?
I intended to preprocessing and compiling the source code first and create cached byte-code of it; then using the `pyc` to run test cases. AFAIK, `gpp` is just a preprocessor, not a compiler.

File: compileFile.js (hunk 3 above)
Comment (mr2rm) [reply compileFile.js]: I wanted to do it myself at first but I couldn't find any way to that except to move it to `outputLocation` after the creation. What's your idea?

Comment (mr2rm) [reply compileFile.js]: The `-b` flag causes bytecode files to be written to their legacy location rather than `__pycache__` in `python3` but `python2` could work without it.

File: extension.js (hunk 7 above)
Comment (mr2rm) [reply extension.js]: Do you mean at the end of the sentence?

File: extension.js (hunk 5 above)
Comment (mr2rm) [reply extension.js]: I'm not agree with throw an error, I think it's better to returns `undefined` which means not found and each module which use it is responsible for it (e.g. show a message about it)

File: compileFile.js (hunk 2 above)
Comment (agrawal-d) [reply compileFile.js]: You're right!

File: compileFile.js (hunk 1 above)
Comment (agrawal-d) [reply compileFile.js]: Yep.

File: extension.js (hunk 5 above)
Comment (mr2rm) [reply extension.js]: I refactored line 428. Let's check it!

File: compileFile.js (hunk 3 above)
Comment (agrawal-d) [reply compileFile.js]: Maybe the comment on the first answer [here](https://stackoverflow.com/questions/13511997/how-to-compile-all-py-file-to-pyc-file-in-a-folder-by-writing-a-python-script/28076884) will work.

File: extension.js (hunk 5 above)
Comment (agrawal-d) [reply extension.js]: Okay.

**Comment by agrawal-d** [comment]
@mr2rm I replied to your comments.

File: locationHelper.js (hunk 9 above)
Comment (mr2rm) [reply locationHelper.js]: Actually it works but I don't think it would be the best practice. Also this option is not available in `compileall` or `py_compile` (actually as command line tool). We can move the generated `.pyc` into `binLocation` after it was created and rename it if we want.

Comment (agrawal-d) [reply locationHelper.js]: Oh! In that case, let's leave it as it is.

**Comment by agrawal-d** [comment]
Let me know when you're done with making the changes so that I can merge this PR.

**Comment by mr2rm** [comment]
At first impression I wanted to separate `python2` and `python3` but I came to that later which It's the better idea to use default version of the host. So it needs to manage the default python version if needed.
But now with these requested changes we have below options to decide about it:

1. Only `python3` will be supported
2. Use the default version of host but check it at runtime and set flags based on the version
3. Separated versions and user can choose any of them (If selected option is not available, a message will be showed and let him to select another one)
4. Separated versions but only available versions will be listed

Any other options? Which one do you prefer?

**Comment by agrawal-d** [comment]
I prefer 1, i.e. Only `python3` will be supported. I don't want us to maintain lots of checking code etc.

**Comment by mr2rm** [comment]
> I prefer 1, i.e. Only `python3` will be supported. I don't want us to maintain lots of checking code etc.

@agrawal-d I'm agree with you and Python 2 will be deprecated soon :smile:

**Comment by agrawal-d** [comment]
Right now we don't do anything if `gcc` or `g++` is not installed. I guess, we can show an info box in the corner if any of the compilation fails due to compiler/interpreter not being installed.

File: locationHelper.js (hunk 9 above)
Comment (mr2rm) [reply locationHelper.js]: Do you mean it's OK to not to store `.pyc` files in `binLocation`?

File: config.js (hunk 4 above)
Comment (mr2rm) [reply config.js]: > Right now we don't do anything if `gcc` or `g++` is not installed. I guess, we can show an info box in the corner if any of the compilation fails due to compiler/interpreter not being installed.

Yeah, I can do that. Also I wrote a snippet for it but according to what you said I think it should be ignored for now .

**Comment by mr2rm** [comment]
@agrawal-d What about to store `.pyc` files in `binLocation`?
If we gonna do it, we have to move it manually after the creation. I couldn't find any other solution to do that. There were any command line flag related to it.
But maybe we can do something for Python 3.2 or above. Let's check [here](https://stackoverflow.com/a/612002/1801351) for more details.

**Comment by mr2rm** [comment]
> I guess, we can show an info box in the corner if any of the compilation fails due to compiler/interpreter not being installed.

Good idea. I'm agree with you again :smile:
Why don't you create a new issue for it?

**Comment by agrawal-d** [comment]
Why are we even generating the bytecode for Python? Let's just skip compilation for python and only run the interpreter to evaluate.

**Comment by mr2rm** [comment]
> Why are we even generating the bytecode for Python? Let's just skip compilation for python and only run the interpreter to evaluate.

We can do it but I preferred to generate bytecode for better performance. Let's skip it...

**Comment by mr2rm** [comment]
> Why are we even generating the bytecode for Python? Let's just skip compilation for python and only run the interpreter to evaluate.

We can do it but I prefer to generate bytecode for better performance. What's wrong with move bytecode to `binLocation` manually?

**Comment by mr2rm** [comment]
@agrawal-d Did you notice my comment?
https://github.com/agrawal-d/competitive-programming-helper/pull/18#issuecomment-605492490

**Comment by agrawal-d** [comment]
>We can do it but I prefer to generate bytecode for better performance. What's wrong with move bytecode to binLocation manually?

There is no performance benefit for executing a bytecode, it only helps in reducing the compilation time when the same file is executed again.[1]

I suggest that you skip the bytecode generation for Python and directly execute it.

[1] See http://effbot.org/pyfaq/can-python-be-compiled-to-machine-code-c-or-some-other-language.htm & https://docs.python.org/3/glossary.html#term-bytecode

**Comment by mr2rm** [comment]
> it only helps in reducing the compilation time when the same file is executed again.

I call it performance, not execution performance; I mean compilation performance.

> Note that the main script executed by Python, even if its filename ends in .py, is not compiled to a .pyc file. It is compiled to bytecode, but the bytecode is not saved to a file. Usually main scripts are quite short, so this doesn’t cost much speed.

I'm agree with above part of the first article which you mentioned. First of all implemented solutions are usually short and also each problem has a few testcases. So bytecode generation cost is not expensive and we can ignore it.

**Comment by agrawal-d** [comment]
@mr2rm are you done with the changes?

**Comment by mr2rm** [comment]
> are you done with the changes?

Yep, It's ready to review. Let's check it!

File: prefrencesHelper.js [hunk 11]
```diff
@@ -1,9 +0,0 @@
-const workspace = require("vscode").workspace;
-
-function preferences() {
-    return workspace.getConfiguration(
-        "competitive-programming-helper"
-    );
-}
-
-module.exports = preferences;
```
Comment (agrawal-d) [inline prefrencesHelper.js]: Why did you delete this file?

File: processManager.js [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+/**
+ * Get rid of created bin if exists
+ * @param {string} filePath bin or .py
+ */
+function removeBin(filePath) {
+    const binPath = getBinLocation(filePath);
+    if (binPath) {
+        spawn("rm", [binPath]);
```
Comment (agrawal-d) [inline processManager.js]: The code here is wrong, as `getBinLocation` for python files returns nothing. Even if it did return the filepath it would be wrong, as wee need `.pyc` path. I've pushed a commit on top to fix this and some other small things.

**Comment by agrawal-d** [comment]
Merged! Thanks for working on this. The refactors you made really improve the quality of the code!

I'll add your name to the README to show your contribution!

File: prefrencesHelper.js (hunk 11 above)
Comment (mr2rm) [reply prefrencesHelper.js]: There were two files `prefrencesHelper.js` and `preferencesHelper.js` which works the same. Also `prefrencesHelper.js` was misspelled and never used in codebase. Evidence showed it was an error so I removed it.

Comment (agrawal-d) [reply prefrencesHelper.js]: Ah cool, thanks!

File: processManager.js (hunk 12 above)
Comment (mr2rm) [reply processManager.js]: It wasn't wrong! The `.pyc`s are created when the file  is imported and running a script create no compiled file. So we need `getBinLocation` returns nothing (`undefined`) for Python.

Comment (agrawal-d) [reply processManager.js]: But `getBinLocation` is not used for Python at all, because you placed checks everywhere to simply return without doing anything; so the only place it used in case of python is for spawning `rm` and `del`.

**Comment by mr2rm** [comment]
> Merged! Thanks for working on this. The refactors you made really improve the quality of the code!

Thanks, I’m glad I could help. :smile: 

> I'll add your name to the README to show your contribution!

It's so kind of you. :pray:  When will you release the new version in VSCode Extensions?

File: processManager.js (hunk 12 above)
Comment (mr2rm) [reply processManager.js]: Yeah, I know. That was kind double-check for safety; Also I wanted `removeBin` works correctly anyway as a function which is responsible for removing generated binary files.

**Comment by agrawal-d** [comment]
> When will you release the new version in VSCode Extensions?

Today!

File: processManager.js (hunk 12 above)
Comment (agrawal-d) [reply processManager.js]: Cool!

**Comment by mr2rm** [comment]
> I'll add your name to the README to show your contribution!

Also don't forget to mention Python in the [Dependencies](https://github.com/agrawal-d/competitive-programming-helper#dependencies).
